nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/drive/PutGoogleDrive.java

```diff
@@ -0,0 +1,495 @@
+
+    private volatile Drive driveService;
+    private Future<File> uploadedFileFuture;
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return RELATIONSHIPS;
+    }
+
+    @Override
+    public List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        String folderId = null;
+        final String filename = context.getProperty(FILE_NAME).evaluateAttributeExpressions(flowFile).getValue();
+        final String mimeType = flowFile.getAttribute(CoreAttributes.MIME_TYPE.key());
+        final String folderName = context.getProperty(FOLDER_NAME).evaluateAttributeExpressions(flowFile).getValue();
+
+        try {
+            folderId = createFoldersAndGetParentId(context, flowFile);
+            final long startNanos = System.nanoTime();
+            final ExecutorService uploadExecutor = Executors.newSingleThreadExecutor();
+            final long size = flowFile.getSize();
+
+            final long chunkUploadThreshold = context.getProperty(CHUNKED_UPLOAD_THRESHOLD)
+                    .asDataSize(DataUnit.B)
+                    .longValue();
+
+            final int uploadChunkSize = context.getProperty(CHUNKED_UPLOAD_SIZE)
+                    .asDataSize(DataUnit.B)
+                    .intValue();
+
+            final String conflictResolution = context.getProperty(CONFLICT_RESOLUTION).getValue();
+
+            Optional<File> alreadyExistingFile = checkFileExistence(filename, folderId);
+            final File fileMetadata = alreadyExistingFile.isPresent() ? alreadyExistingFile.get() : createMetadata(filename, folderId);
+
+            if (alreadyExistingFile.isPresent() && FAIL_RESOLUTION.equals(conflictResolution)) {
+                getLogger().error("File '{}' already exists in folder '{}', conflict resolution is '{}' ", filename, folderId, FAIL_RESOLUTION);
+                flowFile = addAttributes(alreadyExistingFile.get(), flowFile, session);
+                session.transfer(flowFile, REL_FAILURE);
+                return;
+            }
+
+            if (alreadyExistingFile.isPresent() && IGNORE_RESOLUTION.equals(conflictResolution)) {
+                getLogger().info("File '{}' already exists in folder '{}', conflict resolution is '{}' ", filename, folderId, IGNORE_RESOLUTION);
+                flowFile = addAttributes(alreadyExistingFile.get(), flowFile, session);
+                session.transfer(flowFile, REL_SUCCESS);
+                return;
+            }
+
+            final File uploadedFile;
+
+            try (final InputStream rawIn = session.read(flowFile)) {
+                final InputStreamContent mediaContent = new InputStreamContent(mimeType, new BufferedInputStream(rawIn));
```

Comment:
I think we need to close the BufferedInputStream as well. Also, there is another overload of read methods with an `inputStreamCallback`. There is a 50-50 chance of using one, but it adds a lot of logic on top of reading data from a FlowFile. Do we know which one is recommended to use and why? @turcsanyip maybe?